src/handler.ts

```diff
@@ -0,0 +1,685 @@
+        });
+
+        const totalBytes = parseInt(response.headers['content-length']);
+        let receivedBytes = 0;
+
+        response.data.on('data', (data: Buffer | string) => {
+            typeof data === 'string' ? (receivedBytes += Buffer.byteLength(data)) : (receivedBytes += data.length);
+
+            log('download-status', {
+                name: name,
+                type: type,
+                current: receivedBytes,
+                total: totalBytes,
+            });
+        });
+
+        const file = createWriteStream(join(directory, name));
+        response.data.pipe(file);
+
+        await new Promise((resolve) => {
+            file.on('finish', resolve);
+            file.on('error', async (e) => {
+                log('debug', `Failed to download asset to ${join(directory, name)} due to\n${e}. Retrying... ${retry}`);
+                if (existsSync(join(directory, name))) unlinkSync(join(directory, name));
+                if (retry) await downloadAsync(url, directory, name, false, type);
+                return resolve(e);
+            });
+        });
+
+        log('download', name);
+        return {
+            failed: false,
+            asset: null,
+        };
+    } catch (error) {
+        log('debug', `Failed to download asset to ${join(directory, name)} due\n${error}. Retrying... ${retry}`);
+        if (retry) await downloadAsync(url, directory, name, false, type);
+        return;
+    }
+};
+
+const getVersion = async () => {
+    const versionJsonPath = config.versionJson || join(config.directory!, `${config.version.number}.json`);
+    if (existsSync(versionJsonPath)) {
+        parsedVersion = JSON.parse(readFileSync(versionJsonPath).toString());
+        return parsedVersion;
+    }
+
+    const manifest = `${config.url?.meta}/mc/game/version_manifest.json`;
+    const cache = config.cache ? `${config.cache}/json` : `${config.root}/cache/json`;
+    const { data } = await axios.get(manifest);
+
+    if (!existsSync(cache)) {
+        mkdirSync(cache, { recursive: true });
+        log('debug', 'Cache directory created.');
+    }
+
+    writeFileSync(join(cache, 'version_manifest.json'), JSON.stringify(data));
+    log('debug', 'Cached version_manifest.json');
+
+    const desiredVersion = data.versions.find((version: { id: string }) => version.id === config.version.number);
+    if (desiredVersion) {
+        const { data } = await axios.get(desiredVersion.url);
+        writeFileSync(join(`${cache}/${config.version.number}.json`), JSON.stringify(data));
+        log('debug', `Cached ${config.version.number}.json`);
+
+        log('debug', 'Parsed version from version manifest');
+        parsedVersion = data;
+        return parsedVersion;
+    } else {
+        throw Error(`Failed to find version ${config.version.number} in version_manifest.json`);
+    }
+};
+
+const getJar = async () => {
+    await downloadAsync(
+        parsedVersion.downloads.client.url,
+        config.directory!,
```

Comment:
Never use `!`, this assumes you are certain that the type is never undefined (never assume anything). This means either your checks aren't right or your object isn't types properly.